libp2p/connmanager.nim

```diff
@@ -331,35 +345,120 @@ proc storeConn*(c: ConnManager, conn: Connection) =
-    raise newException(CatchableError, "empty peer info")
+    raise newException(CatchableError, "Empty peer info")
+
+  # Launch on close listener
+  # All the errors are handled inside `onClose()` procedure.
+  asyncSpawn c.onClose(conn)
 
   let peerId = conn.peerInfo.peerId
-  if c.conns.getOrDefault(peerId).len > c.maxConns:
-    debug "too many connections",
+  if c.conns.getOrDefault(peerId).len > c.maxPeerConns:
+    debug "Too many connections for peer",
       conn, conns = c.conns.getOrDefault(peerId).len
 
-    raise newTooManyConnections()
+    raise newTooManyConnectionsError()
 
   if peerId notin c.conns:
     c.conns[peerId] = initHashSet[Connection]()
 
   c.conns[peerId].incl(conn)
-
-  # Launch on close listener
-  # All the errors are handled inside `onClose()` procedure.
-  asyncSpawn c.onClose(conn)
   libp2p_peers.set(c.conns.len.int64)
 
   trace "Stored connection",
     conn, direction = $conn.dir, connections = c.conns.len
 
+proc trackConn(c: ConnManager,
+               provider: ConnProvider,
+               dir: Direction):
+               Future[Connection] {.async.} =
+  var conn: Connection
+  try:
+    conn = await provider()
+
+    if isNil(conn):
+      return
+
+    conn.dir = dir
+    trace "Got connection", conn, dir = $dir
+
+    proc semaphoreMonitor() {.async.} =
+      try:
+        await conn.join()
+      except CatchableError as exc:
+        trace "Exception in semaphore monitor, ignoring", exc = exc.msg
+
+      if conn.dir == Direction.In:
+        c.inConnSemaphore.release()
+      else:
+        c.outConnSemaphore.release()
+
+    asyncSpawn semaphoreMonitor()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    if not isNil(conn):
+      await conn.close()
+
+    raise exc
+
+  return conn
+
+proc trackIncomingConn*(c: ConnManager,
+                        provider: ConnProvider):
+                        Future[Connection] {.async.} =
+  ## await for a connection slot before attempting
+  ## to call the connection provider
+  ##
+
+  var conn: Connection
+  try:
+    trace "Tracking incoming connection"
+    await c.inConnSemaphore.acquire()
+    conn = await c.trackConn(provider, Direction.In)
+    if isNil(conn):
+      trace "Couldn't acquire connection, releasing semaphore slot", dir = $Direction.In
+      c.inConnSemaphore.release()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    c.inConnSemaphore.release()
+    raise exc
+
+  return conn
+
+proc trackOutgoingConn*(c: ConnManager,
+                        provider: ConnProvider):
+                        Future[Connection] {.async.} =
+  ## try acquiring a connection if all slots
+  ## are already taken, raise TooManyConnectionsError
+  ## exception
+  ##
+
+  trace "Tracking outgoing connection", count = c.outConnSemaphore.count,
+                                        max = c.outConnSemaphore.size
+
+  if not c.outConnSemaphore.tryAcquire():
```

Comment:
this is a bit odd, as far as limits go: I would typically want to limit the number of total connections and the number of incoming connections, but not the number of outgoing connections.. ie if I want to be connected to 100 peers and manage to make outgoing connections to all of them, there's no need for me to allow incoming connections, but I don't want to wait for incoming connections and _not_ connect to outgoing peers in the meantime.

Reply:
You also want to be able to control the number of outgoing connections, this is a pretty straightforward use case. It can be done in two ways, let the calling code worry about it or allow the library handle this internally. In this case, once connection and peer management becomes more central in libp2p, you want to have a complete view of how both of this resources are being used.

One common use case are bootstrap nodes. Bootstrap nodes have a preference for incoming connections, but might still have outgoing connections to other bootstrap nodes. In this case it makes sense to limit the amount of outgoing connections, you probably want to set 1000 incoming vs 50 outgoing, or something along this lines.

Reply:
but still - there's no way to implement a total peer limit with this setup: if I want 100 peers no matter the direction, the only way to do it becomes limiting both incoming and outgoing to 100, then disconnecting anything outside of that

Reply:
OK, that is a valid point. The implementation assumes limits on both directions, there needs to be a way to have a global limit as well.

For background on this PR. It was never meant to be complete solution to all potential use cases, rather a stopgap that adds the ability to control the amount of incoming outgoing connections. A complete connection management, with proper bidirectional limiting is more involved and nuanced to implement than this simple first pass. 

I would say we have to paths forward with this 

- If connection limiting is something that NBC requires right now - eg, to avoid doing constant disconnects, then I would use this for the time being. With this implementation, we can at least tune the incoming vs outgoing connections in a way that works for now. For example, given that we get a lot less outgoing connections, we could set the proportions to something like 80 incoming vs 20 outgoing.
- If NBC can continue with its current approach of disconnecting peers when it reaches the upper limit, then this PR isn't needed.

I personally think that NBC will benefit from this PR by avoiding the constants disconnects, which add quite a bit of internal churn.

Reply:
yeah, I'm not sure - we simply cannot implement the nbc `maxPeers` option correctly as-is - I suspect the right thing to do for NBC is to continue disconnecting as it does now until total connection limiting is implemented in libp2p, to not impose unnecessary churn on users currently using that option.

Reply:
> not impose unnecessary churn on users currently using that option.

What would the churn for the user be?

Reply:
well, "maxPeers" means total peers in nbc - from a practical point of view, one gets to about 30-40 with outgoing connections only after a day so to get to 160, one would have to set maxPeers to something much higher to account for the new way of measuring and above all, blocking incoming connections

Reply:
Isn't this a heuristic more than the implementation limitation? 

There are two issues at play:

1) NBC has a max number of peers, which is right now implement by disconnecting peers. This is suboptimal, because it requires performing a full connection upgrade and then tearing it down right away - this repeats continuously and it's inefficient.
2) We're getting too little outgoing connections (I suspect this is due to misconfigured NATs), which is an issue with the Eth2 network, not necessarily connection management issue. We have heuristics as to what the proportion of the incoming vs outgoing connection is, it would be up to NBC to pass those ratios correctly to libp2p.

Another approach would be to expose in/out limits as arguments in NBC? (but maybe this is the overhead you want to avoid for the user?)

Reply:
> but still - there's no way to implement a total peer limit with this setup: if I want 100 peers no matter the direction, the only way to do it becomes limiting both incoming and outgoing to 100, then disconnecting anything outside of that

So, actually there is a way of implementing both models fairly simple. Right now I'm using two semaphores one for incoming another for outgoing connections, the simplest way of implementing this would be to just share one semaphore in the case a total count is prefered - that way we can have our cake and eat it too :-)

Reply:
yeah, that would be nice, then we can do it properly in nbc as well

Reply:
with shared semaphore we can do `maxPeers` correctly, but we'll eventually want to have both a global and a directional limit I suspect

Reply:
How is the global limit different from the total directional limit?

Reply:
well, the use case was to set a global limit and a sublimit for either direction (for example max 100 peers of which no more than 80 can be incoming / outgoing)

Reply:
Hmm, ok - but is this indispensable right now? There are numerous sophisticated limits we can think of, but I wonder if this subset will do for now?

As for adding a global limit, we probably need another semaphore.